### PR TITLE
Update: Add optional type-checking, including custom errors

### PR DESCRIPTION
**Pull request by genos**
This PR stems from work we had going internally for something that consumes Quil programs, but it made more sense to upstream it to this repository.

Changes:

- add `src/program/checks.rs`, making it public with a single public function `type_check(&Program) -> ProgramResult<()>`
- add more type-checking-related error types to `src/program/error.rs`
- add `rstest` to the dev dependencies to express a lot of test cases in a rather terse format

### CODE REVIEW

File: src/program/checks.rs [hunk 1]
```diff
@@ -0,0 +1,917 @@
...
+                        data_type_mismatch(instruction, source, "`literal real`", destination, dt)
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (kalzoo) [inline src/program/checks.rs:566]: Excellent tests, thanks!

**Review by kalzoo — APPROVED** [review]
lgtm 🏁

**Comment by kalzoo** [comment]
One thought, for any future reviewers to consider - would it make sense to return a separate error type from the type checks, so that (a) this isn't a breaking change and (b) because these errors can only be returned from `type_check` and are the only errors that can be returned from `type_check`? Then if we want to include them in `ProgramError` later they could be a `TypeError` variant and `thiserror` handles that well.

**Comment by genos** [comment]
> One thought, for any future reviewers to consider - would it make sense to return a separate error type from the type checks, so that (a) this isn't a breaking change and (b) because these errors can only be returned from `type_check` and are the only errors that can be returned from `type_check`? Then if we want to include them in `ProgramError` later they could be a `TypeError` variant and `thiserror` handles that well.

That sounds nice; I'll make that change and revamp the `Breaking:` prefix.

**Comment by genos** [comment]
I've changed the title to try to ensure this doesn't get picked up as a breaking change. Will that be enough for `semantic-release`? Should I modify the commit history here? Squash the commits upon merge? (cc @dbanty)

**Pull request merged**